docs/enhancements/infrastructure-monitoring-service-api/alarms.md

```diff
@@ -0,0 +1,770 @@
+           pk_notification_field TEXT[] DEFAULT ARRAY['alarm_definition_id']::TEXT[],
+           created_at TIMESTAMPTZ DEFAULT CURRENT_TIMESTAMP,
+           CONSTRAINT fk_alarm_definitions_version FOREIGN KEY (alarm_dictionary_version) REFERENCES versions(version_number),
+           CONSTRAINT unique_alarm_name_last_change UNIQUE (alarm_name, alarm_last_change)
+);
+
+-- Table: probable_causes
+DROP TABLE IF EXISTS probable_causes CASCADE;
+CREATE TABLE probable_causes (
+             probable_cause_id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+             alarm_definition_id UUID UNIQUE,
+             FOREIGN KEY (alarm_definition_id) REFERENCES alarm_definitions(alarm_definition_id) ON DELETE CASCADE
+);
+
+-- Create a new entry for probable cause for each new alarm_definition_id
+CREATE OR REPLACE FUNCTION insert_probable_cause()
+    RETURNS TRIGGER AS $$
+BEGIN
+    -- Insert a new row into probable_causes with the alarm_definition_id from the new alarm_definition
+    INSERT INTO probable_causes (alarm_definition_id)
+    VALUES (NEW.alarm_definition_id);
+
+    RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER trg_insert_probable_cause
+    AFTER INSERT ON alarm_definitions
+    FOR EACH ROW
+EXECUTE FUNCTION insert_probable_cause();
+
+
+-- Table: alarm_subscription_info
+DROP TABLE IF EXISTS alarm_subscription_info CASCADE;
+CREATE TABLE alarm_subscription_info (
+             subscription_id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+             consumer_subscription_id UUID,
+             filter TEXT,
+             callback TEXT NOT NULL,
+             largest_number_alarm_event_seen_so_far BIGINT NOT NULL DEFAULT 0,
+             created_at TIMESTAMPTZ DEFAULT CURRENT_TIMESTAMP
+);
+
+
+CREATE TYPE Status AS ENUM ('resolved', 'firing');
+CREATE TYPE PerceivedSeverity AS ENUM ('CRITICAL', 'MAJOR', 'MINOR','WARNING', 'INDETERMINATE', 'CLEARED');
+
+-- SEQUENCE: Counter to keep track of latest events and use it notify only the latest
+CREATE SEQUENCE alarm_sequence_seq
+    START WITH 1
+    INCREMENT BY 1
+    NO MINVALUE
+    NO MAXVALUE
+    CACHE 1;
+
+-- Table: alarm_event_record
+DROP TABLE IF EXISTS alarm_event_record CASCADE;
+CREATE TABLE alarm_event_record (
+            alarm_event_record_id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+            alarm_definition_id UUID NOT NULL,
+            probable_cause_id UUID,
+            status Status DEFAULT 'firing' NOT NULL,
+            alarm_raised_time TIMESTAMPTZ NOT NULL,
+            alarm_changed_time TIMESTAMPTZ,
+            alarm_cleared_time TIMESTAMPTZ,
+            alarm_acknowledged_time TIMESTAMPTZ,
+            alarm_acknowledged BOOLEAN NOT NULL default FALSE,
+            perceived_severity PerceivedSeverity NOT NULL,
+            extensions JSONB,
+            created_at TIMESTAMPTZ DEFAULT CURRENT_TIMESTAMP,
+            finger_print TEXT not null,
+            alarm_sequence_number BIGINT DEFAULT nextval('alarm_sequence_seq'),
```

Comment:
I presume you intend on using this sequence number to control the order that these records are presented to a subscription, but I believe this is going to be problematic for events that may change states multiple times before we get to send it up to a subscriber.   If a record changes multiple times it may be possible that the subscriber will only see the last update.  To address this it may be necessary to maintain a separate table just to track the individual transitions of each alarm_event_record so that subscribers see the full picture.  

Maybe I've misread this, but this came to mind when I saw the trigger function below that updates the sequence value on an update.